torchao/dtypes/nf4tensor.py

```diff
@@ -1,20 +1,26 @@
 import functools
 from dataclasses import dataclass
 from typing import Dict, Tuple
+import math
+import sys
+from enum import Enum, auto
 
 import torch
 import torch.nn.functional as F
 from torch import Tensor
+from torch.distributed.device_mesh import DeviceMesh
+from torch._prims_common import make_contiguous_strides_for
 
 
 aten = torch.ops.aten
 
 c10d_functional = torch.ops.c10d_functional
 
-from typing import Any, Tuple
+from typing import Any, Optional, Tuple, Union, List
 
 NF4_OPS_TABLE: Dict[Any, Any] = {}
 
+INNER_TENSOR_NAMES_FOR_SHARDING = ["quantized_scalers", "quantization_factor", "quantized_data"]
```

Comment:
So this is something FSDP2 requires any Tensor subclass to have defined?

Reply:
for any Tensor subclass, we prefer reusing `__tensor_flatten__` to lookup inner tensors. For NF4, we define INNER_TENSOR_NAMES_FOR_SHARDING as a subset of `__tensor_flatten__` because `scaler_mean` and `nf4` are too tiny to shard

Reply:
Hm, isn't that something you could filter with a `numel` based heuristic within FSDP itself instead of requiring some tensor subclasses to communicate it?

Reply:
I think the inner tensors that are sharded needs to match the `torch.chunk` implementation in the subclass, so FSDP cannot necessarily determine the tensors to shard itself. (E.g., if FSDP filtered by numel but the subclass implemented `torch.chunk` to still shard some tensor smaller than the numel threshold, then there would be a correctness issue.)

Reply:
changed to private const with underscore `_INNER_TENSOR_NAMES_FOR_SHARDING` after discussion with @cpuhrsch